zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2508,11 +2509,17 @@ public void connect(final long maxMsToWaitUntilConnected)
    */
   public void connect(final long maxMsToWaitUntilConnected, Watcher watcher)
       throws ZkInterruptedException, ZkTimeoutException, IllegalStateException {
+    acquireEventLock();
+
     if (isClosed()) {
       throw new IllegalStateException("ZkClient already closed!");
     }
+    if (_currentState != null) {
+      throw new IllegalStateException(
+          "ZkClient is not in init state. connect() has already been called.");
+    }
+
```

Comment:
Why do we need the lock at the beginning?

Reply:
`_currentState` and `_isClosed` are set in critical section in `close()`. 
We need to check these 2 state and create connection in critical section as well.